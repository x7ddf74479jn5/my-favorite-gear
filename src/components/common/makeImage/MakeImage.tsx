import Grid from "@mui/material/Grid";
import { styled } from "@mui/system";
import type { VFC } from "react";
import React, { memo } from "react";

import type { FavoriteList } from "@/services/models/favoriteList";
import type { Gear } from "@/services/models/gear";

const Root = styled("div")(({ theme }) => ({
  display: "flex",
  flexWrap: "wrap",
  justifyContent: "space-around",
  margin: theme.spacing(2, 0),
}));

interface Tile {
  src: string;
  alt: string;
}
const GridImage: VFC<{ favoriteList?: FavoriteList; gear?: Gear }> = ({
  favoriteList,
  gear,
}) => {
  if (favoriteList) {
    const tile: Tile[] = favoriteList.gears.map((v) => {
      return {
        src: v.mediumImageUrl || "",
        alt: v.productName,
      };
    });
    tile.splice(4, 0, {
      src: favoriteList.image?.replace("_normal", "") || "",
      alt: favoriteList.twitterId,
    });

    return (
      <>
        <Grid
          container
          columns={3}
          spacing={0}
          sx={{ width: 360, height: 360 }}
        >
          {tile.map((item: Tile, index) => {
            return (
              <Grid item key={`${item.src}-${index}`} columns={1}>
                <img src={item.src} alt={item.alt} />
              </Grid>
            );
          })}
        </Grid>
      </>
    );
  }

  return (
    <>
      <Grid container columns={3} spacing={0} sx={{ width: 360, height: 360 }}>
        <Grid item columns={3}>
          <img src={gear?.mediumImageUrl || ""} alt={gear?.productName || ""} />
        </Grid>
      </Grid>
    </>
  );
};

const MakeImage: VFC<{ favoriteList?: FavoriteList; gear?: Gear }> = ({
  favoriteList,
  gear,
}) => {
  return (
    <>
      <Root>
        <GridImage favoriteList={favoriteList} gear={gear} />
      </Root>
    </>
  );
};

export default memo(MakeImage);
