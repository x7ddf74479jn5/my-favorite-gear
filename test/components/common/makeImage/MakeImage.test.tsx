import MakeImage from "components/common/makeImage/MakeImage";
import React from "react";

import { render, testFavoriteList, testGear } from "../../../test-utils";

describe("MakeImage", () => {
  it("matches snapshot", () => {
    let renderResult = render(
      <MakeImage favoriteList={testFavoriteList} gear={undefined} />
    );
    expect(renderResult.asFragment()).toMatchSnapshot();

    renderResult = render(
      <MakeImage favoriteList={undefined} gear={testGear} />
    );
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("renders 3x3 tiles when provided only favoriteList", () => {
    const renderResult = render(
      <MakeImage favoriteList={testFavoriteList} gear={undefined} />
    );

    const listElement = renderResult.getByRole("list");
    expect(listElement).toBeInTheDocument();
    expect(listElement).toHaveAttribute(
      "class",
      expect.stringContaining("gridlist")
    );
    expect(renderResult.getAllByRole("listitem")).toHaveLength(9);
    expect(renderResult.getAllByRole("listitem")[0]).toHaveStyle({
      height: "120px",
    });
    const twitterTileImg = renderResult.getAllByRole("img")[4];
    expect(twitterTileImg).toHaveAttribute("src", "image");
    expect(twitterTileImg).toHaveAttribute("alt", "twitterId");
  });

  it("renders 1 tile when provided only gear", () => {
    const renderResult = render(
      <MakeImage favoriteList={undefined} gear={testGear} />
    );

    const listElement = renderResult.getByRole("list");
    expect(listElement).toBeInTheDocument();
    expect(listElement).toHaveAttribute(
      "class",
      expect.stringContaining("gridlist")
    );
    expect(renderResult.getAllByRole("listitem")).toHaveLength(1);
    expect(renderResult.getAllByRole("img")).toHaveLength(1);
    const imageElement = renderResult.getByRole("img");
    expect(imageElement).toHaveAttribute("src", "src");
    expect(imageElement).toHaveAttribute("alt", testGear.productName);
  });

  it("renders none when provided neither favoriteList nor gear", () => {
    const renderResult = render(
      <MakeImage favoriteList={undefined} gear={undefined} />
    );

    const rootElement = renderResult.container.firstChild;
    expect(rootElement).toHaveAttribute(
      "class",
      expect.stringContaining("root")
    );
    const listElement = renderResult.getByRole("list");
    expect(listElement).toBeInTheDocument();
    expect(listElement).toHaveAttribute(
      "class",
      expect.stringContaining("gridlist")
    );
    expect(renderResult.getAllByRole("listitem")).toHaveLength(1);
  });
});
