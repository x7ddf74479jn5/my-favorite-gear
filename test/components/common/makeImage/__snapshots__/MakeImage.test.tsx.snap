// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MakeImage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <ul
      class="MuiGridList-root makeStyles-gridlist-2"
      style="margin: 0px;"
    >
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName1"
            src="src1"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName2"
            src="src2"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName3"
            src="src3"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName4"
            src="src4"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="twitterId"
            src="image"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName5"
            src="src5"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName6"
            src="src6"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName7"
            src="src7"
          />
        </div>
      </li>
      <li
        class="MuiGridListTile-root"
        style="width: 33.333333333333336%; height: 120px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName8"
            src="src8"
          />
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`MakeImage matches snapshot 2`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <ul
      class="MuiGridList-root makeStyles-gridlist-2"
      style="margin: 0px;"
    >
      <li
        class="MuiGridListTile-root"
        style="width: 100%; height: 360px; padding: 0px;"
      >
        <div
          class="MuiGridListTile-tile"
        >
          <img
            alt="productName"
            src="src"
          />
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
